src/ir/ir_to_proto_test.cc

```diff
@@ -0,0 +1,131 @@
+#include "google/protobuf/util/message_differencer.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/ir_printer.h"
+#include "src/ir/proto/raksha_ir.pb.h"
+#include "src/ir/proto_to_ir.h"
+#include "src/parser/ir/ir_parser.h"
+
+namespace raksha::ir {
+
+class IrToProtoTest : public testing::TestWithParam<std::string_view> {};
+
+TEST_P(IrToProtoTest, RoundTripPreservesAllInformation) {
+  const auto input_program_text = GetParam();
+  IrProgramParser ir_parser;
+  // TODO: IRPrinter cannot take a const SsaNames.
+  /*const*/ auto parse_result = ir_parser.ParseProgram(input_program_text);
+  const Module& program = *parse_result.module;
+  const IRContext& context1 = *parse_result.context;
+  /*const*/ auto& ssa_names = *parse_result.ssa_names;
+
+  // Check round trip toString (sanity check to ensure later checks are sound).
+  EXPECT_EQ(IRPrinter::ToString(program, ssa_names), input_program_text);
+
+  proto::IrTranslationUnit orig_translation_unit_proto =
+      IRToProto::Convert(context1, program);
+
+  IRContext context2;
+  ProtoToIR::Result recovered = ProtoToIR::Convert(context2, orig_translation_unit_proto);
+  const std::string result = IRPrinter::ToString(*recovered.module, *recovered.ssa_names);
+  EXPECT_EQ(result, input_program_text);
+
+  proto::IrTranslationUnit recovered_translation_unit_proto =
+      IRToProto::Convert(context1, *recovered.module);
+
+  google::protobuf::util::MessageDifferencer differencer;
+  std::string difference = "<Compare not yet run>";
+  differencer.ReportDifferencesToString(&difference);
+  ASSERT_TRUE(differencer.Compare(orig_translation_unit_proto,
+                                  recovered_translation_unit_proto))
+      << "RECOVERED DATA DID NOT MATCH!\n"
+      << difference;
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    IrToProtoTest, IrToProtoTest,
+    ::testing::Values(
+        R"(module m0 {
+  block b0 {
+    %0 = core.plus []()
+  }  // block b0
+  block b1 {
+    %1 = core.plus []()
+  }  // block b1
+}  // module m0
+)",
+        R"(module m0 {
+  block b0 {
+    %0 = core.plus []()
+  }  // block b0
+  block b1 {
+    %1 = core.plus [access: "private", transform: "no"](<<ANY>>, <<ANY>>)
+  }  // block b1
+}  // module m0
+)",
+        R"(module m0 {
+  block b0 {
+    %0 = core.plus []()
+  }  // block b0
+  block b1 {
+    %1 = core.plus [access: "private", transform: "no"](%0, <<ANY>>)
+  }  // block b1
+}  // module m0
+)",
+/*
+// TODO: Support out of order declarations.
+R"(module m0 {
+  block b0 {
+    %0 = core.plus [access: "private", transform: "no"](%3, <<ANY>>)
+    %1 = core.mult [access: "private", transform: "no"](%3, %0)
+  }  // block b0
+  block b1 {
+    %2 = core.merge [](<<ANY>>, <<ANY>>)
+    %3 = core.select []()
+  }  // block b1
+}  // module m0
+)",
+*/
+R"(module m0 {
+  block b0 {
+    %0 = core.select []()
+    %1 = core.merge [](<<ANY>>, <<ANY>>)
+  }  // block b0
+  block b1 {
+    %2 = core.plus [access: "private", transform: "no"](%0, <<ANY>>)
```

Comment:
Heads up. Moving forward we are planning on not supporting value access between blocks, instead have blocks return values to be used in other blocks (Exact details need to be figured out). In that case, this test would fail as `%0` in block b1 would be initialized in block b0. Please let us know your thoughts about this!

Reply:
Thanks, Harsha. Just to add a little more context. We should have the inputs and outputs of a block to be more constrained. Allowing operations in one block to directly access operations in another block will make it hard to reason about behavior of blocks. 

Harsha and I were chatting about this and we decided that the names should be unique within the block alone. If we want to support a block from accessing the result of another block, we might want to have a different syntax with a block prefix like `b0.%1`.

Reply:
Gotcha. Thanks for the heads up. Can we land this and iterate?